### PR TITLE
Add delete defaults option

### PR DESCRIPTION
**Pull request by jeffwatkins**
If you set something in UserDefaults the easiest way to reset it is to delete it when the test is complete.

https://github.com/classpass/Pterodactyl/assets/169875/7dd1a7ab-98bd-412b-89ee-184bfcb7ee29

### CODE REVIEW

File: PterodactylExample/PterodactylExampleUITests/PterodactylExampleUITests.swift [hunk 1]
```diff
@@ -35,6 +35,32 @@ class PterodactylExampleUITests: XCTestCase {
         XCTAssertEqual(defaultsValue.label, testValue.uuidString)
     }
 
+    func testDeleteUserDefaults() throws {
+        let pterodactyl = Pterodactyl(targetAppBundleId: "com.mattstanford.PterodactylExample")
+        let testValue = UUID()
+        pterodactyl.updateDefaults([
+            "Test": .string(testValue.uuidString)
+        ])
```
Comment (jeffwatkins) [inline PterodactylExample/PterodactylExampleUITests/PterodactylExampleUITests.swift:43]: Before we start the app, we add a userDefaults value for `Test` that is the string value of a UUID.

File: PterodactylExample/PterodactylExampleUITests/PterodactylExampleUITests.swift [hunk 2]
```diff
@@ -35,6 +35,32 @@ class PterodactylExampleUITests: XCTestCase {
...
+
+        app.launch()
+
+        waitForElementToAppear(object: app.staticTexts["Pterodactyl Example"])
+
+        let defaultsValue = app.staticTexts["DefaultsValue"].firstMatch
+        XCTAssertTrue(defaultsValue.exists)
+        XCTAssertEqual(defaultsValue.label, testValue.uuidString)
```
Comment (jeffwatkins) [inline PterodactylExample/PterodactylExampleUITests/PterodactylExampleUITests.swift:51]: Ensure the app is able to read the user defaults value and that it matches.

File: PterodactylExample/PterodactylExampleUITests/PterodactylExampleUITests.swift [hunk 3]
```diff
@@ -35,6 +35,32 @@ class PterodactylExampleUITests: XCTestCase {
...
+
+        app.terminate()
+
+        pterodactyl.deleteDefaults(for: ["Test"])
```
Comment (jeffwatkins) [inline PterodactylExample/PterodactylExampleUITests/PterodactylExampleUITests.swift:55]: After the app terminates, delete the user defaults value for `Test` and relaunch the app.

File: PterodactylExample/PterodactylExampleUITests/PterodactylExampleUITests.swift [hunk 4]
```diff
@@ -35,6 +35,32 @@ class PterodactylExampleUITests: XCTestCase {
...
+        app.launch()
+
+        waitForElementToAppear(object: app.staticTexts["Pterodactyl Example"])
+
+        XCTAssertTrue(defaultsValue.exists)
+        XCTAssertEqual(defaultsValue.label, "MISSING VALUE")
```
Comment (jeffwatkins) [inline PterodactylExample/PterodactylExampleUITests/PterodactylExampleUITests.swift:61]: The app should be unable to find the user defaults value and therefore displays "MISSING VALUE" instead.

File: PterodactylLib/Built_Product/PterodactylServer.app/Contents/Info.plist
Comment (jeffwatkins) [inline PterodactylLib/Built_Product/PterodactylServer.app/Contents/Info.plist:1]: This is part of the binary server app shipped with the library. It's safe to ignore it.

File: PterodactylLib/Built_Product/PterodactylServer.app/Contents/MacOS/PterodactylServer
Comment (jeffwatkins) [inline PterodactylLib/Built_Product/PterodactylServer.app/Contents/MacOS/PterodactylServer:1]: This is part of the binary server app shipped with the library. It's safe to ignore it.

File: PterodactylLib/Built_Product/PterodactylServer.app/Contents/MacOS/PterodactylServer.debug.dylib
Comment (jeffwatkins) [inline PterodactylLib/Built_Product/PterodactylServer.app/Contents/MacOS/PterodactylServer.debug.dylib:1]: This is part of the binary server app shipped with the library. It's safe to ignore it.

File: PterodactylLib/Built_Product/PterodactylServer.app/Contents/MacOS/__preview.dylib
Comment (jeffwatkins) [inline PterodactylLib/Built_Product/PterodactylServer.app/Contents/MacOS/__preview.dylib:1]: This is part of the binary server app shipped with the library. It's safe to ignore it.

File: PterodactylLib/Built_Product/PterodactylServer.app/Contents/Resources/Assets.car
Comment (jeffwatkins) [inline PterodactylLib/Built_Product/PterodactylServer.app/Contents/Resources/Assets.car:1]: This is part of the binary server app shipped with the library. It's safe to ignore it.

File: PterodactylLib/Built_Product/PterodactylServer.app/Contents/Resources/Base.lproj/Main.storyboardc/MainMenu.nib
Comment (jeffwatkins) [inline PterodactylLib/Built_Product/PterodactylServer.app/Contents/Resources/Base.lproj/Main.storyboardc/MainMenu.nib:1]: This is part of the binary server app shipped with the library. It's safe to ignore it.

File: PterodactylLib/Built_Product/PterodactylServer.app/Contents/_CodeSignature/CodeResources
Comment (jeffwatkins) [inline PterodactylLib/Built_Product/PterodactylServer.app/Contents/_CodeSignature/CodeResources:1]: This is part of the binary server app shipped with the library. It's safe to ignore it.

File: PterodactylLib/Classes/Endpoint.swift [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+//  Copyright © 2024 Matt Stanford. All rights reserved.
+//
+
+import Foundation
+
+enum Endpoint: String, CaseIterable {
+    case push = "simulatorPush"
+    case updateDefaults
+    case deleteDefaults
+}
```
Comment (jeffwatkins) [inline PterodactylLib/Classes/Endpoint.swift:14]: This abstracts out the endpoints used in both the server & library.

File: PterodactylLib/Classes/Pterodactyl.swift [hunk 6]
```diff
@@ -93,6 +82,39 @@ public class Pterodactyl {
...
+    }
+
+    public func withDefaults(_ defaults: [String: UpdateDefaultsValue], block: (() throws -> Void)) rethrows {
+        updateDefaults(defaults)
+        defer {
+            // Ensure we delete the defaults values even if the block we called throws an error.
+            deleteDefaults(for: Array(defaults.keys))
+        }
+        try block()
+    }
```
Comment (jeffwatkins) [inline PterodactylLib/Classes/Pterodactyl.swift:112]: This is how most code should use the user defaults:
```
pterodactyl.withDefaults(["Test": 100, "monkeys": "banana"]) {
…
}
```
Before executing the block the user defaults will be set and after executing the block, the user defaults will be cleared.

File: PterodactylLib/Classes/Requests.swift [hunk 7]
```diff
@@ -26,3 +26,9 @@ struct UpdateDefaultsRequest: Codable {
     let appBundleId: String
     let defaults: [String: UpdateDefaultsValue]
 }
+
+struct DeleteDefaultsRequest: Codable {
+    let simulatorId: String
+    let appBundleId: String
+    let keys: [String]
+}
```
Comment (jeffwatkins) [inline PterodactylLib/Classes/Requests.swift:34]: This is the structure sent between the library and the server for deleting user defaults keys.

**Review by irobles25 — APPROVED** [review]

**Review by kelsinator — APPROVED** [review]
looking forward to being able to set feature flags with this!

**Pull request merged**